File: src/java/nz/co/searchwellington/controllers/ProfileController.java
package nz.co.searchwellington.controllers;

import java.io.IOException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import nz.co.searchwellington.model.SiteInformation;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.repositories.ConfigRepository;
import nz.co.searchwellington.repositories.ResourceRepository;
import nz.co.searchwellington.statistics.StatsTracking;

import org.apache.log4j.Logger;
import org.springframework.web.servlet.ModelAndView;


public class ProfileController extends BaseMultiActionController {
    
    Logger log = Logger.getLogger(ProfileController.class);
    
    private SiteInformation siteInformation;
     
    public ProfileController(ResourceRepository resourceDAO, UrlStack urlStack, ConfigRepository configDAO, SiteInformation siteInformation) {       
        this.resourceDAO = resourceDAO;
        this.urlStack = urlStack;
        this.configDAO = configDAO;
        this.siteInformation = siteInformation;
    }
    
       
    public ModelAndView profile(HttpServletRequest request, HttpServletResponse response) throws IOException {
        ModelAndView mv = new ModelAndView();
        
        urlStack.setUrlStack(request);
        User loggedInUser = setLoginState(request, mv);
        StatsTracking.setRecordPageImpression(mv, configDAO.getStatsTracking());        
                        
        mv.addObject("heading", "Profile");
        populateSecondaryLatestNewsitems(mv, loggedInUser);
        
        mv.setViewName("profile");
        return mv;
    }


        
   
       
    

}
    

File: src/main/scala/nz/co/searchwellington/controllers/ProfileController.scala
package nz.co.searchwellington.controllers

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.repositories.ContentRetrievalService
import nz.co.searchwellington.repositories.mongo.MongoRepository
import nz.co.searchwellington.urls.UrlBuilder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.servlet.ModelAndView

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}
import scala.jdk.CollectionConverters._

@Order(3)
@Controller
class ProfileController @Autowired()(mongoRepository: MongoRepository, loggedInUserFilter: LoggedInUserFilter,
                                     val contentRetrievalService: ContentRetrievalService)
  extends ReasonableWaits with CommonModelObjects {

  @GetMapping(Array("/profile/edit")) def edit(): ModelAndView = {
    val loggedInUser = loggedInUserFilter.getLoggedInUser
    Await.result(for {
      commonLocal <- commonLocal
    } yield {
      new ModelAndView("editProfile").
        addObject("heading", "Editing your profile").
        addObject("user", loggedInUser).addAllObjects(commonLocal)
    }, TenSeconds)
  }

  /*
  // TODO reinstate
  @PostMapping(Array("/profile/edit")) def save(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    loggedInUserFilter.getLoggedInUser.map { loggedInUser =>
      if (request.getParameter("profilename") != null && Await.result(isValidAvailableProfilename(request.getParameter("profilename")), TenSeconds)) {
        //  loggedInUser.setProfilename(request.getParameter("profilename"))
      }
      //  loggedInUser.setName(request.getParameter("name"))
      // loggedInUser.setBio(request.getParameter("bio"))
      //   loggedInUser.setUrl(request.getParameter("url"))
      mongoRepository.saveUser(loggedInUser)
      new ModelAndView(new RedirectView(urlBuilder.getProfileUrlFromProfileName(loggedInUser.getProfilename)))
    }.getOrElse {
      null //TODO
    }
  }
  */

  def isValidAvailableProfilename(profilename: String): Future[Boolean] = {
    for {
      existingUser <- mongoRepository.getUserByProfilename(profilename)
    } yield {
      profilename.matches("[a-z|A-Z|0-9]+") && existingUser.isEmpty
    }
  }

}